Add template program

tadeboro opened the pull request:
Newly added program can be used to customize application configuration
at runtime through environment variables.

Mode of operation is similar to how bash interpolates variables in
strings, main difference being that this utility works on file
contents.

Commit:
Add template program

Newly added program can be used to customize application configuration
at runtime through environment variables.

Mode of operation is similar to how bash interpolates variables in
strings, main difference being that this utility works on file
contents.

gberginc commented on template.c:
```diff
@@ -0,0 +1,75 @@
+
+	if (argc != 3)
+	{
+		fputs ("Exactly two arguments expected.\n", stderr);
```
I siggest you name the args. Eg, source and dest file

tadeboro replied:
Will do.

miha-plesko commented on template.c:
```diff
@@ -0,0 +1,75 @@
+{
+	char *tmpl = NULL;
+	char buffer[1024];
+	size_t n = 0;
```
Nitpicking, but I think we could hardcode this in L67

tadeboro replied:
We cannot, because getline() takes pointer to size as a second argument.

miha-plesko reviewed (approved):
@tadeboro thanks, this looks great to me. I'm a bit rusty in C, but from what I understand this should work.

@justinc1 can you please review and merge?

gberginc commented on template.c:
```diff
@@ -0,0 +1,75 @@
+{
+	for (;;) {
+		char *start = strstr (tmpl, "${");
+		if (! start)
```
Suggest to use {} even for one line ifs

tadeboro replied:
Will fix this.

gberginc commented on template.c:
```diff
@@ -0,0 +1,75 @@
+		if (! end)
+			break;
+
+		*start = '\0';
```
Null terminal at start? Is it necessary?

tadeboro replied:
`start` points to the start of the template variable. Anything preceeding it should be copied verbatim to the output. And to avoid doing pointer acrobatics, I simply placed `NULL` at the end of the verbatim text and used `strcpy` to copy it later.

gberginc commented on template.c:
```diff
@@ -0,0 +1,75 @@
+		expand_template (tmpl, buffer);
+		fputs (buffer, out);
+	}
+	fclose (out);
```
Should in also be closed?

tadeboro replied:
Input has beed depleted at this point, so closing it is a bit pointless (unless we care about symetry with output). Output is closed explicitly in order to flush the buffers to the disk.

gberginc reviewed (changes requested):
Few minor comments and requests for clarifications/changes.

justinc1 commented on README.md:
```diff
@@ -3,3 +3,23 @@
+should point to the destination file. For example:
+
+    $ cat a.txt
+    Name of master is ${MASTER_NAME} and worker is ${WORKER_NAME}.
```
Mention that $MASTER_NAME is not expanded, only ${MASTER_NAME}.

tadeboro replied:
I should probably specify template language. I will add a section about what walid templates are.

justinc1 commented on template.c:
```diff
@@ -0,0 +1,75 @@
+		if (! envvar)
+			envvar = "";
+
+		strcpy (result, tmpl);
```
strncpy

tadeboro replied:
I know, I know, but this will complicate things a bit;)

justinc1 commented on template.c:
```diff
@@ -0,0 +1,75 @@
+	}
+
+	/* Copy template part after the last variable. */
+	strcpy (result, tmpl);
```
strncpy

tadeboro commented:
I refactored the application in concordance with the comments. All unsafe function calls (strcpy) have been removed and replaced by memcpy calls with proper bound checking.

I decided against the use of strncpy, because I expect that vast majority of the template files in the wild will have short lines and filling the buffer with trailing zeros makes little sense.

I also added a documentation about the algorithm that is being used to expand the template and renamed a few variables to have more meaningful names.

gberginc commented on template.c:
```diff
@@ -0,0 +1,114 @@
+
+		*variable_end = '\0';
+		char const *envvar = getenv(variable_start);
+		if (! envvar)
```
May I ask for this as well?

gberginc reviewed (commented):
Apart from the one additional `if` without `{}`, your changes are fantastic!

Thanks @tadeboro!

tadeboro commented:
I added missing braces to the `if` clause.

gberginc reviewed (approved):
Thanks @tadeboro. And apologies for nitpicking 😄 .
